### PR TITLE
Spiralogram: Reimplementation in pyqtgraph

### PR DESCRIPTION
**Pull request by janezd**
##### Issue

Spiralogram depends on highcharts, and could also be improved in other ways.

##### Description of changes

A new implementation in pure Qt.

**To do**:

- [x] Refactor
- [x] Contexts (including, e.g. pending selection)
- [x] Tests
- [x] Copy `VariableBinner` to utilities in core
- [x] Check whether old aggregations (and possibly some other code) in `agg_funcs` is still relevant; remove if not.
- [x] Time on radial axis
- [x] Remove local file from preview
- [x] ~~Format legend with g instead of with f?~~ Doesn't help - it's already g :)

##### Includes
- [X] Code changes
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by ajdapretnar** [comment]
@janezd Would you mind rebasing? It seems some old commits got their way into this one.

**Comment by ajdapretnar** [comment]
A few comments:
- Radial would benefit from having the time periods as it used to. I found it very useful. See documentation for a simple use-case.
- The legend seems to force the plot to be overly small or big. I propose to have the legend as an overlay, as with other visualization widgets where you can move it around. Probably smart formatting could help.
Screenshot 2022-07-08 at 10 07 36
- I get assertion error when clicking, for example, Radial=Adj Close (Yahoo Timeseries) and then changing Time Period to Day of Year.

```
-------------------------- AssertionError Exception ---------------------------
Traceback (most recent call last):
  File "/Users/ajda/orange/orange-widget-base/orangewidget/gui.py", line 2251, in __call__
    self.func(**kwds)
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 538, in _x_var_changed
    self.reblock()
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 674, in reblock
    self.recompute(nocommit=nocommit)
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 681, in recompute
    self.redraw()
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 693, in redraw
    self.draw_segments()
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 883, in draw_segments
    segment = SegmentItem.from_coordinates(
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 125, in from_coordinates
    assert x < nperiods
AssertionError
-------------------------------------------------------------------------------
```

- Comboboxes fall out of view when attributes are selected.
Screenshot 2022-07-08 at 10 06 29
- I would prefer having Selected Data as the first possible output and Statistics as the second. I think it would be more in line with other widgets.

**Comment by ajdapretnar** [comment]
Connecting Yahoo Finance and then changing the data set to Iris (replacing YF with File), causes:
```
---------------------------- ValueError Exception -----------------------------
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 674, in reblock
    self.recompute(nocommit=nocommit)
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 678, in recompute
    self.computed_data = self.compute_data() if self.block_data else None
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 776, in compute_data
    color_data = self.data.get_column_view(self.color_var)[0]
  File "/Users/ajda/orange/orange3/Orange/data/table.py", line 1597, in get_column_view
    col_index = self.domain.index(index)
  File "/Users/ajda/orange/orange3/Orange/data/domain.py", line 345, in index
    raise ValueError("'%s' is not in domain" % var)
ValueError: 'Low' is not in domain
-------------------------------------------------------------------------------
```

**Comment by ajdapretnar** [comment]
The widget crashes on sparse data:
```
---------------------------- ValueError Exception -----------------------------
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 673, in reblock
    self.block_data = self.compute_block_data() if self.data else None
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 736, in compute_block_data
    {(x, 0): np.flatnonzero(x_data == x)
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 736, in <dictcomp>
    {(x, 0): np.flatnonzero(x_data == x)
  File "<__array_function__ internals>", line 180, in flatnonzero
  File "/Users/ajda/.pyenv-x86/versions/py3.9/lib/python3.9/site-packages/numpy/core/numeric.py", line 663, in flatnonzero
    return np.nonzero(np.ravel(a))[0]
  File "<__array_function__ internals>", line 180, in nonzero
  File "/Users/ajda/.pyenv-x86/versions/py3.9/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 1958, in nonzero
    return _wrapfunc(a, 'nonzero')
  File "/Users/ajda/.pyenv-x86/versions/py3.9/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
  File "/Users/ajda/.pyenv-x86/versions/py3.9/lib/python3.9/site-packages/scipy/sparse/_base.py", line 337, in __bool__
    raise ValueError("The truth value of an array with more than one "
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all().
-------------------------------------------------------------------------------
```

**Comment by janezd** [comment]
> Radial would benefit from having the time periods as it used to. I found it very useful. See documentation for a simple use-case.

I started working on it, then I realized that we should probably solve this problem in general.

"Radial" is not different from any other splitting of data by categorical values. If you observe time series in Box plot or Distributions, wouldn't you find it useful to split/group data by years, too? Maybe we need a widget that would get time series and add variables "Year", "Month", ... and so forth; the widget would be sufficiently intelligent to see which variables make sense, given the time span.

Such widget would not allow only the functionality that I removed from Spiralogram, but also add it to all other widgets, including those in core.

What do you think?

**Comment by janezd** [comment]
> The legend seems to force the plot to be overly small or big. 

The screenshot is weird, this shouldn't have happened. The chart shouldn't fall outside the widget. Something is apparently wrong when the legend does not fit the widget. I'll investigate.

The values in the legend are unfortunate, though. I wouldn't care too much about such cases.

I can of course make the legend movable, but it wouldn't help. The way it works is that the widget computes the maximal size for the circle, such that it wouldn't overlap with the legend. If the widget ignored the legend in computation of the size, the legend would not fit anywhere, so moving it would not help. (There is some slack due to labels around the circle, but not much.) Considering the displays are landscape, making the widget wide enough should suffice to fit the legend and the circle. When I fix the problem that you discovered.

**Comment by janezd** [comment]
> I get assertion error when clicking, for example, Radial=Adj Close (Yahoo Timeseries) and then changing Time Period to Day of Year.

Fixed.

**Comment by janezd** [comment]
> Having a single instance, then clicking on a plot to select the only slice there is, causes:

When fixing this, I ran into the following dilemma: if your x-axis is month of year and radially you have something else (say gender), but not all combinations occur - what would you like to see in output "Statistics"?

Before fixing the bug, you'd see only months for which there is some data, but those months would list all genders, including, say, women in May though there may be no women in May. I changed this so that you only get rows with non-zero counts. This makes sense because aggregations for such rows would always be nan. But perhaps we'd like to see counts for all combinations of months and genders, including zero counts?

**Comment by janezd** [comment]
> The widget crashes on sparse data:
> But this is probably the least important issue on the list and could be solved in a separate PR.

I'd prefer it this way.

**Comment by janezd** [comment]
> - I would prefer having Selected Data as the first possible output and Statistics as the second. I think it would be more in line with other widgets.

Done.

**Comment by janezd** [comment]
Thanks for quick reviewing before I forgot how the code is organized. :)

I've split your comments into multiple replies for easier follow ups.

**Comment by janezd** [comment]
I now also resolved the problem with the legend making the graph too large. It happened when the legend width exceeded half of the widget's width, which would occur if the widget is really small or a product of values is too large. Now the widget ignores the legend if it's that wide; the user can then either scroll or extend the widget.

I do not care too much because using product in such cases is unreasonable anyway.

**Comment by ajdapretnar** [comment]
> "Radial" is not different from any other splitting of data by categorical values. If you observe time series in Box plot or Distributions, wouldn't you find it useful to split/group data by years, too? Maybe we need a widget that would get time series and add variables "Year", "Month", ... and so forth; the widget would be sufficiently intelligent to see which variables make sense, given the time span.

- You are right. However, from the user perspective, Timeseries add-on is where I'd expect it most. I wonder how much simplicity in Spiralogram we are losing with gaining additional options in other widgets.

- The product issue (large legend values) remains. I cannot even see the entire periodogram, even if I expand the window across the entire screen. I agree, the use case is stupid. But is there perhaps a way to place the legend below the visualiation? Or perhaps show rounded values in the legend (1k-5k, 54.6 * 10^8.... or something)? Just an idea, I don't insist on fixing this, because it is really not very useful (the product, not the nice legend).

- Now I get an AssertionError just by clicking Day of month on Yahoo Finance.
```
-------------------------- AssertionError Exception ---------------------------
Traceback (most recent call last):
  File "/Users/ajda/orange/orange-widget-base/orangewidget/gui.py", line 2251, in __call__
    self.func(**kwds)
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 567, in _x_var_changed
    self.reblock()
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 703, in reblock
    self.recompute(nocommit=nocommit)
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 710, in recompute
    self.redraw()
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 722, in redraw
    self.draw_segments()
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 919, in draw_segments
    segment = SegmentItem.from_coordinates(
  File "/Users/ajda/orange/orange3-timeseries/orangecontrib/timeseries/widgets/owspiralogram.py", line 125, in from_coordinates
    assert x < nperiods
AssertionError
-------------------------------------------------------------------------------
```

**Comment by ajdapretnar** [comment]
> I can't reproduce this one and don't see how it could have occurred. Could these be some settings from an earlier crash?

Yes, I think this is the case. I tried to crash the widget with cleared settings, but I didn't manage. So I guess we let this go for now.

File: orangecontrib/timeseries/widgets/owspiralogram.py [hunk 1]
```diff
@@ -277,170 +462,609 @@ class Inputs:
...
+            rows = np.hstack([seg_indices[coord] for coord in self.selection])
+            data = self.data[rows]
+        self.Outputs.selected_data.send(data)
 
 
 if __name__ == "__main__":
-    WidgetPreview(OWSpiralogram).run(Table.from_file('airpassengers'))
+    WidgetPreview(OWSpiralogram).run(
+        Timeseries.from_file(
+            '/Users/janez/Downloads/slovenia-traffic-accidents-2016-events.tab')
```
Comment (ajdapretnar) [inline orangecontrib/timeseries/widgets/owspiralogram.py:1069]: I am sorry, but this will not pass my review. 😉

Comment (janezd) [reply orangecontrib/timeseries/widgets/owspiralogram.py]: Well, I tried. :)

**Comment by janezd** [comment]
**Regarding product:** Let the useless legend let the user know (s)he's doing something stupid.

**Re time on radial axis:** I again started implementing it. Then realized that I shouldn't just throw everything into the combo: if your time interval spans years, you don't want to let the user split data by seconds. So I wanted to fill the combo only with units that would split the data into reasonable number of bins. And, finally, I realized that the widget already has it. If you want to see data split radially by years, just select the time variable and set the binning to 1 year. The code behind binning lets you choose exactly the intervals that are sensible All I needed to improve was to show years, not year intervals, e.g. 2018 instead of 2018 - 2019.

Screen Shot 2022-07-11 at 22 25 29

In this fashion you cannot split radially by. for instance, Month of year. And I don't care: this should go around the circle, not radially.

Preview no longer loads a local file.

This may perhaps be it.

**Comment by ajdapretnar** [comment]
Thanks! This is fantastic! I'll update the documentation in a single batch after all the three (four?) widgets are migrated to PyQt.

**Pull request merged**